Add both Ursa compiler targets to UPP regression tests

SamuelTrahanNOAA opened the pull request:
## Description

Port to Ursa.

Changes:

1. Usage message; fixes #1236 
2. Correct test names in change log; fixes #1237 
1. Ursa regression tests for intel and intelllvm which share a common `jobs-dev/*URSA.sh`
1. Add compiler name (`_intel` or `_intelllvm`) to rt.log filenames
1. Ursa has two baselines: `data_out_intel` and `data_out_intelllvm`

Note: The libIFI and GTG aren't ported to Ursa in this PR. If we're going to port those to Ursa, it'll be in a separate PR.

### Status

- [x] Compile with older intel or intelllvm
- [x] rt.sh can run either compiler
- [x] Ursa tests pass with older intel
- [x] Ursa tests pass with intelllvm

## Relevant issues

- fixes #1236
- fixes #1237 

## Support for two compilers

The new `-C` (capital C) option to rt.sh chooses the compiler. Mandatory on Ursa!

```bash
rt.sh -C [intel|intelllvm]
```

On other platforms, it'll default to intel.

Also on Ursa, there are:

- `data_out_intel` and `data_out_intelllvm`
- `runtime.log.URSA_intel` and `runtime.log.URSA_intelllvm`

There's only one set of jobs-dev files on Ursa; they work for either compiler

## `rt.sh` Argument List Error Detection

In testing, we found people would often make mistakes in the rt.sh command line. Bash's simple `getopts` built-in command cannot detect most common errors.

I've added three safety checks to argument parsing, on top of the one getopts already has.

*If an argument begins with a `-` (like `rt.sh -r -C intel`) the script will abort due to a missing argument.*

<details> <summary>Sample Output</summary>

```
./rt.sh -r -C intel
+ SECONDS=0
+ git_branch=develop
+ git_url=https://github.com/NOAA-EMC/UPP.git
+ clone_on=no
+ disable_ifi=no
+ print_full_help=no
+ build_exe=yes
+ compiler=MISSING
+ set +x
Argument error: -r requires an argument

Usage: rt.sh -a account -C compiler -r /path/to/scrub/space [-options] [compiler]
Executes UPP regression tests. Includes IFI tests if ../sorc/libIFI.fd exists.

Results are here:
  ../tests/logs/MACHINE_compiler.log = report of regression tests for each machine and compiler.
  changed_results.txt = A list of tests whose results have changed.

Always set these:
  -a account = accounting code for job submission. Default account is often overused. Always set this!
  -C = chosen compiler. Default: intel. Mandatory on Ursa!
  -r rundir = path to a scrub space. Default area is often over quota. Always set this!

General options:
  -d = disable ifi tests even if ifi is available
  -e = don't build the UPP executable
  -h homedir = path to the regression test data
  -w workdir = directory to store per-job batch and log files.
  -H = print full help message and exit. Includes special-use options.

------------------------------------------------------------------------
FATAL ERROR: Script is exiting due to a missing argument. See error message above
------------------------------------------------------------------------
```

</details>

*If an argument is empty (like `rt.sh -r '' -C intel`) the script will abort due to an empty argument.*

<details> <summary>Sample Output</summary>

```
./rt.sh -r '' -C intel
+ SECONDS=0
+ git_branch=develop
+ git_url=https://github.com/NOAA-EMC/UPP.git
+ clone_on=no
+ disable_ifi=no
+ print_full_help=no
+ build_exe=yes
+ compiler=MISSING
+ set +x
Argument error: -r argument is the empty string

Usage: rt.sh -a account -C compiler -r /path/to/scrub/space [-options] [compiler]
Executes UPP regression tests. Includes IFI tests if ../sorc/libIFI.fd exists.

Results are here:
  ../tests/logs/MACHINE_compiler.log = report of regression tests for each machine and compiler.
  changed_results.txt = A list of tests whose results have changed.

Always set these:
  -a account = accounting code for job submission. Default account is often overused. Always set this!
  -C = chosen compiler. Default: intel. Mandatory on Ursa!
  -r rundir = path to a scrub space. Default area is often over quota. Always set this!

General options:
  -d = disable ifi tests even if ifi is available
  -e = don't build the UPP executable
  -h homedir = path to the regression test data
  -w workdir = directory to store per-job batch and log files.
  -H = print full help message and exit. Includes special-use options.

------------------------------------------------------------------------
FATAL ERROR: Script is exiting due to invalid argument. See error message above.
------------------------------------------------------------------------
```

</details>

*Detecting spurious or positional arguments (like `rt.sh -r /path /another/path -C intel intelllvm` exits with an error.*

<details> <summary>Sample Output</summary>

```
./rt.sh -r /some/path /another/path -C intel intellvm
+ SECONDS=0
+ git_branch=develop
+ git_url=https://github.com/NOAA-EMC/UPP.git
+ clone_on=no
+ disable_ifi=no
+ print_full_help=no
+ build_exe=yes
+ compiler=MISSING
+ set +x

Usage: rt.sh -a account -C compiler -r /path/to/scrub/space [-options] [compiler]
Executes UPP regression tests. Includes IFI tests if ../sorc/libIFI.fd exists.

Results are here:
  ../tests/logs/MACHINE_compiler.log = report of regression tests for each machine and compiler.
  changed_results.txt = A list of tests whose results have changed.

Always set these:
  -a account = accounting code for job submission. Default account is often overused. Always set this!
  -C = chosen compiler. (Capital C) Default: intel. Mandatory on Ursa!
  -r rundir = path to a scrub space. Default area is often over quota. Always set this!

General options:
  -d = disable ifi tests even if ifi is available
  -e = don't build the UPP executable
  -h homedir = path to the regression test data
  -w workdir = directory to store per-job batch and log files.
  -H = print full help message and exit. Includes special-use options.

------------------------------------------------------------------------
FATAL ERROR: Script is aborting due to spurious arguments: /another/path -C intel intellvm
------------------------------------------------------------------------
```

</details>

*getopts already detects unknown -options*

<details> <summary>Sample Output</summary>

```
./rt.sh -W -H -A -T
+ SECONDS=0
+ git_branch=develop
+ git_url=https://github.com/NOAA-EMC/UPP.git
+ clone_on=no
+ disable_ifi=no
+ print_full_help=no
+ build_exe=yes
+ compiler=MISSING
+ set +x
./rt.sh: illegal option -- W

Usage: rt.sh -a account -C compiler -r /path/to/scrub/space [-options] [compiler]
Executes UPP regression tests. Includes IFI tests if ../sorc/libIFI.fd exists.

Results are here:
  ../tests/logs/MACHINE_compiler.log = report of regression tests for each machine and compiler.
  changed_results.txt = A list of tests whose results have changed.

Always set these:
  -a account = accounting code for job submission. Default account is often overused. Always set this!
  -C = chosen compiler. Default: intel. Mandatory on Ursa!
  -r rundir = path to a scrub space. Default area is often over quota. Always set this!

General options:
  -d = disable ifi tests even if ifi is available
  -e = don't build the UPP executable
  -h homedir = path to the regression test data
  -w workdir = directory to store per-job batch and log files.
  -H = print full help message and exit. Includes special-use options.

------------------------------------------------------------------------
FATAL ERROR: Invalid -option. See error message above.
------------------------------------------------------------------------
```

</details>

SamuelTrahanNOAA committed:
port to ursa intel llvm (called "intel" in modulefiles and scripts)

SamuelTrahanNOAA committed:
use the same KPO list for fv3gefs pe and non-pe tests on HERA

BenjaminBlake-NOAA commented:
@SamuelTrahanNOAA Thank you for working on this.  We are actually about to get rid of all of the pe tests (including fv3gefs_pe) with PR #1208 .  So from my perspective I think it is okay that the fv3gefs and fv3gefs_pe results do not match.  @WenMeng-NOAA what do you think?

WenMeng-NOAA commented:
> @SamuelTrahanNOAA Thank you for working on this. We are actually about to get rid of all of the pe tests (including fv3gefs_pe) with PR #1208 . So from my perspective I think it is okay that the fv3gefs and fv3gefs_pe results do not match. @WenMeng-NOAA what do you think?

I agree with Ben's comments and also suggest moving all updates related to the UPP regression tests on Ursa into a separate PR, so we can commit the changes for Ursa more quickly.

SamuelTrahanNOAA committed:
compiler is in all modulefile names and is an argument to rt.sh (mandatory on ursa)

SamuelTrahanNOAA committed:
longer wallclock for 3drtm

SamuelTrahanNOAA committed:
all ci jobs are exclusive on ursa

SamuelTrahanNOAA committed:
ursa older intel compiler modulefile

SamuelTrahanNOAA committed:
put $compiler in runtime.log in rt.sh

SamuelTrahanNOAA commented:
> I agree with Ben's comments and also suggest moving all updates related to the UPP regression tests on Ursa into a separate PR, so we can commit the changes for Ursa more quickly.

@WenMeng-NOAA - There are changes to the build system and modulefile naming to properly support two compilers.

DavidHuber-NOAA reviewed (approved):
Approving the module file updates (I can't speak to the scripts).  Thanks for separating them clearly into `intel` and `intelllvm`.

SamuelTrahanNOAA committed:
revert accidental change: upp_common.lua should have no _intel

SamuelTrahanNOAA committed:
put compiler name in rt.log filename

SamuelTrahanNOAA committed:
hera tests passed

SamuelTrahanNOAA committed:
ursa tests passed

SamuelTrahanNOAA committed:
Merge branch 'feature/ursa' of ssh://github.com/SamuelTrahanNOAA/UPP into feature/ursa

SamuelTrahanNOAA commented:
The baselines on Orion and Hercules are not up to date. They're for a hash `3e89d51c20108c29943fa7183eb17d99815a1012` which isn't anywhere in NOAA-EMC/UPP repository.

SamuelTrahanNOAA committed:
merge upstream develop

SamuelTrahanNOAA committed:
merge upstream develop

SamuelTrahanNOAA committed:
remove pe tests from ursa

SamuelTrahanNOAA committed:
merge upstream develop

SamuelTrahanNOAA committed:
merge compile_upp.sh changes from other porting branch

SamuelTrahanNOAA commented:
I've merged develop to this PR, so the latest workflow changes are all here.

Also, I synced the changes from the other ursa-porting PR.

SamuelTrahanNOAA committed:
merge upstream develop

SamuelTrahanNOAA committed:
ursa tests passed again

WenMeng-NOAA commented on ci/jobs-dev/run_post_hrrr_URSA.sh:
```diff
@@ -0,0 +1,129 @@
+ # if not bit-identical, use cmp_grib2_grib2 to compare each grib record
+ export err1=$?
+ if [ $err1 -eq 0 ] ; then
+  msg="hrrr test: your new post executable generates bit-identical ${filein2} as the trunk"
```
@SamuelTrahanNOAA Please replace the string 'the trunk' with 'the develop branch' in all ursa job cards.

SamuelTrahanNOAA replied:
Done.

SamuelTrahanNOAA committed:
the trunk => the develop branch

SamuelTrahanNOAA committed:
rt.sh: fix an option parsing glitch and add a usage message

SamuelTrahanNOAA committed:
merge develop

SamuelTrahanNOAA committed:
move compiler to -C option

SamuelTrahanNOAA committed:
remove spurious []

SamuelTrahanNOAA committed:
rt.sh: add -C option

SamuelTrahanNOAA committed:
compiler="MISSING"

SamuelTrahanNOAA committed:
Merge branch 'feature/ursa' of ssh://github.com/SamuelTrahanNOAA/UPP into feature/ursa

SamuelTrahanNOAA committed:
rt-status.py: add the missing .split()

SamuelTrahanNOAA committed:
more fixes to error detection

SamuelTrahanNOAA committed:
merge upstream

WenMeng-NOAA commented:
@SamuelTrahanNOAA Thanks for your contribution on porting the UPP RTs on Ursa. Could you add a change log with your name in rt.sh?

SamuelTrahanNOAA committed:
Ursa tests need compiler to be exported

SamuelTrahanNOAA committed:
author info

SamuelTrahanNOAA commented:
The Hercules and Orion tests are reporting differences in two files. Nothing in this PR should be able to change the results. Are the baselines up to date?

```
There are changes in results for case hrrr in /work2/noaa/wrfruc/strahan/upp/orion-feature-ursa/ci/rundir/upp-ORION/hrrr_2020060118/WRFTWO.GrbF04
There are changes in results for case rap in /work2/noaa/wrfruc/strahan/upp/orion-feature-ursa/ci/rundir/upp-ORION/rap_2020072316/WRFPRS.GrbF16
```

The diff for hrrr is empty, but the diff for rap looks like this:

```
707:194080921:CDCON:convective cloud layer:rpn_corr=-nan:rpn_rms=undefined
```

SamuelTrahanNOAA commented:
I'm seeing the same difference report on Hera, too. I haven't run on WCOSS yet.

```
There are changes in results for case hrrr in /scratch1/NCEPDEV/stmp2/Samuel.Trahan/upp-HERA/hrrr_2020060118/WRFTWO.GrbF04
There are changes in results for case rap in /scratch1/NCEPDEV/stmp2/Samuel.Trahan/upp-HERA/rap_2020072316/WRFPRS.GrbF16
```

SamuelTrahanNOAA commented:
I've retested and fixed a few bugs. The tests are reporting changes to rap and hrrr; see above.

1. rt-status.py correctly prints the test name
2. compiler is now in a `-C intel|intelllvm` (capital C) option instead of a positional argument
3. authorship log update (requested by @WenMeng-NOAA )
4. usage message in rt.sh when detecting argument errors, or if `-H` is given
5. `rt.sh -H` prints a usage message and exits without running regression tests

SamuelTrahanNOAA committed:
ursa tests passed

SamuelTrahanNOAA committed:
remove redundant assignment to runtime_log (overwritten a few lines down)

WenMeng-NOAA commented:
@SamuelTrahanNOAA rt.sh is not correctly handling with "rt.sh  -r <rundir>".

SamuelTrahanNOAA commented:
> rt.sh is not correctly handling with "rt.sh -r ".

@WenMeng-NOAA - What behavior are you seeing?

The -r option requires an argument. If you don't give it one, the next argument is assumed to be the rundir. If the -r is at the end of the argument list, you should see "./rt.sh: option requires an argument -- r"

SamuelTrahanNOAA committed:
clarify usage message

SamuelTrahanNOAA committed:
rt.sh: clarify usage message

SamuelTrahanNOAA committed:
Wen Meng is author of cmp_grib2_grib2.sh

SamuelTrahanNOAA commented:
My Ursa scripts were not comparing grib2 files because $cmp_grib2_grib2 wasn't set on that platform. I'm working on that problem now...

SamuelTrahanNOAA committed:
set cmp_grib2_grib2 in rt.sh

SamuelTrahanNOAA commented:
Crisis averted. The $cmp_grib2_grib2 is now set in rt.sh and passed to other jobs through the environment.

SamuelTrahanNOAA committed:
Merge branch 'feature/ursa' of ssh://github.com/SamuelTrahanNOAA/UPP into feature/ursa

SamuelTrahanNOAA committed:
fix typo

SamuelTrahanNOAA committed:
Merge branch 'feature/ursa' of ssh://github.com/SamuelTrahanNOAA/UPP into feature/ursa

SamuelTrahanNOAA committed:
ursa tests passed

WenMeng-NOAA commented:
> > rt.sh is not correctly handling with "rt.sh -r ".
> 
> @WenMeng-NOAA - What behavior are you seeing?
> 
> The -r option requires an argument. If you don't give it one, the next argument is assumed to be the rundir. If the -r is at the end of the argument list, you should see "./rt.sh: option requires an argument -- r"

@SamuelTrahanNOAA I am using "rt.sh -r $my_rundir" on Ursa. But the value of $my_rundir isn't being passed to $rundir.

WenMeng-NOAA commented:
@SamuelTrahanNOAA Can you syn your PR with the latest UPP develop for our testing? At this time, there is no need to commit the RT logs.

SamuelTrahanNOAA committed:
Merge remote-tracking branch 'upstream/develop' into feature/ursa

SamuelTrahanNOAA commented:
> I am using "rt.sh -r $my_rundir" on Ursa. But the value of $my_rundir isn't being passed to $rundir.

@WenMeng-NOAA -

That would fail if $my_rundir was empty. I just tried the -r option and it worked for me.

To debug this, I need:
1. The exact command you sent to rt.sh
2. Values of any environment variables that are in the command
3. The stdout and stderr streams

You can send this to me by email if you prefer the entire world not know your directory paths.

SamuelTrahanNOAA committed:
rt.sh: check for empty arguments and arguments that begin with -

SamuelTrahanNOAA commented:
I've added two safety checks to the dash option parsing:

*If an argument begins with a `-` (like `rt.sh -r -C intel`) the script will abort due to a missing argument.*

<details> <summary>Sample Output</summary>

```
./rt.sh -r -C intel
+ SECONDS=0
+ git_branch=develop
+ git_url=https://github.com/NOAA-EMC/UPP.git
+ clone_on=no
+ disable_ifi=no
+ print_full_help=no
+ build_exe=yes
+ compiler=MISSING
+ set +x
Argument error: -r requires an argument

Usage: rt.sh -a account -C compiler -r /path/to/scrub/space [-options] [compiler]
Executes UPP regression tests. Includes IFI tests if ../sorc/libIFI.fd exists.

Results are here:
  ../tests/logs/MACHINE_compiler.log = report of regression tests for each machine and compiler.
  changed_results.txt = A list of tests whose results have changed.

Always set these:
  -a account = accounting code for job submission. Default account is often overused. Always set this!
  -C = chosen compiler. Default: intel. Mandatory on Ursa!
  -r rundir = path to a scrub space. Default area is often over quota. Always set this!

General options:
  -d = disable ifi tests even if ifi is available
  -e = don't build the UPP executable
  -h homedir = path to the regression test data
  -w workdir = directory to store per-job batch and log files.
  -H = print full help message and exit. Includes special-use options.

------------------------------------------------------------------------
FATAL ERROR: Script is exiting due to a missing argument. See error message above
------------------------------------------------------------------------
```

</details>


*If an argument is empty (like `rt.sh -r '' -C intel`) the script will abort due to an empty argument.*

<details> <summary>Sample Output</summary>

```
./rt.sh -r '' -C intel
+ SECONDS=0
+ git_branch=develop
+ git_url=https://github.com/NOAA-EMC/UPP.git
+ clone_on=no
+ disable_ifi=no
+ print_full_help=no
+ build_exe=yes
+ compiler=MISSING
+ set +x
Argument error: -r argument is the empty string

Usage: rt.sh -a account -C compiler -r /path/to/scrub/space [-options] [compiler]
Executes UPP regression tests. Includes IFI tests if ../sorc/libIFI.fd exists.

Results are here:
  ../tests/logs/MACHINE_compiler.log = report of regression tests for each machine and compiler.
  changed_results.txt = A list of tests whose results have changed.

Always set these:
  -a account = accounting code for job submission. Default account is often overused. Always set this!
  -C = chosen compiler. Default: intel. Mandatory on Ursa!
  -r rundir = path to a scrub space. Default area is often over quota. Always set this!

General options:
  -d = disable ifi tests even if ifi is available
  -e = don't build the UPP executable
  -h homedir = path to the regression test data
  -w workdir = directory to store per-job batch and log files.
  -H = print full help message and exit. Includes special-use options.

------------------------------------------------------------------------
FATAL ERROR: Script is exiting due to invalid argument. See error message above.
------------------------------------------------------------------------
```

</details>

*It already knew how to detect invalid options*

<details> <summary>Sample Output</summary>

```
./rt.sh -W -H -A -T
+ SECONDS=0
+ git_branch=develop
+ git_url=https://github.com/NOAA-EMC/UPP.git
+ clone_on=no
+ disable_ifi=no
+ print_full_help=no
+ build_exe=yes
+ compiler=MISSING
+ set +x
./rt.sh: illegal option -- W

Usage: rt.sh -a account -C compiler -r /path/to/scrub/space [-options] [compiler]
Executes UPP regression tests. Includes IFI tests if ../sorc/libIFI.fd exists.

Results are here:
  ../tests/logs/MACHINE_compiler.log = report of regression tests for each machine and compiler.
  changed_results.txt = A list of tests whose results have changed.

Always set these:
  -a account = accounting code for job submission. Default account is often overused. Always set this!
  -C = chosen compiler. Default: intel. Mandatory on Ursa!
  -r rundir = path to a scrub space. Default area is often over quota. Always set this!

General options:
  -d = disable ifi tests even if ifi is available
  -e = don't build the UPP executable
  -h homedir = path to the regression test data
  -w workdir = directory to store per-job batch and log files.
  -H = print full help message and exit. Includes special-use options.

------------------------------------------------------------------------
FATAL ERROR: Invalid -option. See error message above.
------------------------------------------------------------------------
```

</details>

SamuelTrahanNOAA committed:
move "fatal error" message to bottom of usage() output

SamuelTrahanNOAA committed:
Synopsis -> Usage

SamuelTrahanNOAA committed:
clarify which -c/-C is lower/upper case (some people have bad terminal fonts)

SamuelTrahanNOAA committed:
detect positional arguments

SamuelTrahanNOAA committed:
remove broken positional argument detection; will redo shortly

SamuelTrahanNOAA committed:
rt.sh: detect positional arguments and exit if they are present

SamuelTrahanNOAA committed:
rt.sh: remove unneeded debug echo

SamuelTrahanNOAA commented:
I've added a critical safety check to the rt.sh argument parsing: detect positional arguments. That should find spurious arguments.

<details> <summary>Sample Output</summary>

```
./rt.sh -r /some/path /another/path -C intel intellvm
+ SECONDS=0
+ git_branch=develop
+ git_url=https://github.com/NOAA-EMC/UPP.git
+ clone_on=no
+ disable_ifi=no
+ print_full_help=no
+ build_exe=yes
+ compiler=MISSING
+ set +x

Usage: rt.sh -a account -C compiler -r /path/to/scrub/space [-options] [compiler]
Executes UPP regression tests. Includes IFI tests if ../sorc/libIFI.fd exists.

Results are here:
  ../tests/logs/MACHINE_compiler.log = report of regression tests for each machine and compiler.
  changed_results.txt = A list of tests whose results have changed.

Always set these:
  -a account = accounting code for job submission. Default account is often overused. Always set this!
  -C = chosen compiler. (Capital C) Default: intel. Mandatory on Ursa!
  -r rundir = path to a scrub space. Default area is often over quota. Always set this!

General options:
  -d = disable ifi tests even if ifi is available
  -e = don't build the UPP executable
  -h homedir = path to the regression test data
  -w workdir = directory to store per-job batch and log files.
  -H = print full help message and exit. Includes special-use options.

------------------------------------------------------------------------
FATAL ERROR: Script is aborting due to spurious arguments: /another/path -C intel intellvm
------------------------------------------------------------------------
```

</details>

SamuelTrahanNOAA committed:
remove parsed arguments from rt.sh command line

BenjaminBlake-NOAA commented:
@SamuelTrahanNOAA Thank you for the improvements to rt.sh.  I tested a few of the various added options and they work well for me.

For my tests on WCOSS2 and Hera, I got "no changes in test results detected."  But when testing on Ursa, I was able to reproduce your changed results in the rap and hrrr tests, where the diff file for HRRR is blank but the diff file for RAP shows an entry for CDCON.  The rap and hrrr baselines were just modified with the merging of PR #1227 so I wonder if the baseline was not updated on Ursa.

WenMeng-NOAA commented:
@SamuelTrahanNOAA Sine you included the script 'cmp_grib2_grib2.sh' under ci/, I suggest adding the line to load the wgrib2 module in all job cards under jobs-dev/.

BenjaminBlake-NOAA commented:
@SamuelTrahanNOAA I backed out the recent updates to params_grib2_tbl_new which cause the baseline changes, and my RTs on Ursa completed successfully with no baseline changes.  Please update your rap and hrrr baseline data on Ursa - thanks.

SamuelTrahanNOAA commented:
> Sine you included the script 'cmp_grib2_grib2.sh' under ci/, I suggest adding the line to load the wgrib2 module in all job cards under jobs-dev/.

@WenMeng-NOAA - Let's do one test at a time. I'll update the Ursa baseline and you two can confirm that works. Then, once that's done, I'll update all systems and you can confirm that works too.

I didn't update all systems before because the amount of testing is quite large. You'll have to check each machine to confirm the latest wgrib2 is able to detect changes.

WenMeng-NOAA commented:
> > Sine you included the script 'cmp_grib2_grib2.sh' under ci/, I suggest adding the line to load the wgrib2 module in all job cards under jobs-dev/.
> 
> @WenMeng-NOAA - Let's do one test at a time. I'll update the Ursa baseline and you two can confirm that works. Then, once that's done, I'll update all systems and you can confirm that works too.
> 
> I didn't update all systems before because the amount of testing is quite large. You'll have to check each machine to confirm the latest wgrib2 is able to detect changes.

@SamuelTrahanNOAA Your plan sounds good to me.

SamuelTrahanNOAA commented:
>  I backed out the recent updates to params_grib2_tbl_new which cause the baseline changes, and my RTs on Ursa completed successfully with no baseline changes. Please update your rap and hrrr baseline data on Ursa - thanks.

@BenjaminBlake-NOAA - Can I achieve this by merging develop to my branch and regenerating the baselines? Or am I misunderstanding?

BenjaminBlake-NOAA commented:
@SamuelTrahanNOAA Yes that should work, thanks

SamuelTrahanNOAA committed:
ursa tests passed again

SamuelTrahanNOAA committed:
use stmp for default rundir path on ursa

SamuelTrahanNOAA commented:
@BenjaminBlake-NOAA - I've updated the Ursa baselines. Please test them and confirm they work for you.

Can you think of an easy way to change the results for all tests? I need to confirm that wgrib2 can detect changes. When I changed -O3 to -O2 in the `sorc/ncep_post.fd/CMakeFile` it didn't change the results.

SamuelTrahanNOAA committed:
clarify an error message

SamuelTrahanNOAA committed:
Merge branch 'feature/ursa' of ssh://github.com/SamuelTrahanNOAA/UPP into feature/ursa

BenjaminBlake-NOAA commented:
@SamuelTrahanNOAA Sounds good, I am testing on Ursa now.

I have an idea for a quick way to change the results for all tests - I'll let you know if it works.

SamuelTrahanNOAA commented:
Changing T on one point in all ranks changed the results. The collapsed region below contains the diff.

<details> <summary>Diff of changes</summary>

```diff
diff --git a/sorc/ncep_post.fd/PROCESS.f b/sorc/ncep_post.fd/PROCESS.f
index e10a5bfa..b3c0c497 100644
--- a/sorc/ncep_post.fd/PROCESS.f
+++ b/sorc/ncep_post.fd/PROCESS.f
@@ -44,7 +44,9 @@
       use upp_ifi_mod, only: run_ifi
       use CTLBLK_mod, only: cfld, etafld2_tim, eta2p_tim, mdl2sigma_tim, surfce2_tim,&
                             mdl2agl_tim, mdl2thandpv_tim, calrad_wcloud_tim,&
-                            cldrad_tim, miscln_tim, fixed_tim, ntlfld, me, run_ifi_tim
+                            cldrad_tim, miscln_tim, fixed_tim, ntlfld, me, run_ifi_tim, &
+                            ista, jsta
+      use VRBLS3D, only: T
 !- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
       implicit none
 !
@@ -63,6 +65,14 @@
 !****************************************************************************
 !     START SUBROUTINE PROCESS.
 !
+
+
+!****************************************************************************
+!     FIXME: DO NOT MERGE THIS!!!
+      if(me==0) write(*,*) "Modify T on all ranks to ensure results change"
+      T(ISTA,JSTA,3) = T(ISTA,JSTA,3) * 1.01
+!****************************************************************************
+
       cfld=0
       if(me==0) write(*,*) "PROCESS starts"
 !

```

</details>

BenjaminBlake-NOAA commented:
Yeah that works.  My suggestion was going to be to rename any variable that is used by all the tests inside parm/params_grib2_tbl_new, and that does change results.

My intel and intelllvm tests both completed successfully on Ursa with no baseline changes.

SamuelTrahanNOAA committed:
use ci/cmp_grib2_grib2 on all platforms

SamuelTrahanNOAA committed:
load wgrib2 on hera

SamuelTrahanNOAA commented:
@BenjaminBlake-NOAA - Can you update the scripts for wcoss2 to use `ci/cmp_grib2_grib2.sh`? I've almost finished Orion, Hercules, and Hera.

BenjaminBlake-NOAA commented:
@SamuelTrahanNOAA Sure.  It sounds like I just need to add "module load wgrib2/2.0.8" to all the scripts since you updated rt.sh to use ci/cmp_grib2_grib2.sh?

SamuelTrahanNOAA commented:
I'm using wgrib2/3.1.1 on the RDHPCS machines because that is what comes with spack stack

SamuelTrahanNOAA committed:
use cmp_grib2_grib2.sh  on orion

SamuelTrahanNOAA committed:
Merge branch 'feature/ursa' of ssh://github.com/SamuelTrahanNOAA/UPP into feature/ursa

SamuelTrahanNOAA committed:
use cmp_grib2_grib2.sh  on hercules

SamuelTrahanNOAA commented:
@BenjaminBlake-NOAA - I've updated the branch. All four supported RDHPCS machines use the spack stack wgrib2 and ci/cmp_grib2_grib2.sh. That's Hera, Hercules, Orion, and Ursa.

BenjaminBlake-NOAA commented:
@WenMeng-NOAA it looks like you use wgrib2/2.0.8 on WCOSS2 and I also don't see wgrib2/3.1.1 as an official installation.  So I updated the WCOSS2 scripts to use wgrib2/2.0.8 but that can be changed if needed.

Commit:
use cmp_grib2_grib2.sh on WCOSS2

BenjaminBlake-NOAA commented:
@SamuelTrahanNOAA Thanks for updating all the RDHPCS machines.  I am leaving the office soon but I can test these changes tomorrow morning.

SamuelTrahanNOAA commented:
Mismatched wgrib2 versions probably won't cause a problem. It's something to keep in mind though, if we encounter bugs or need to add fields the older wgrib2 doesn't recognize.

WenMeng-NOAA commented:
> Yeah that works. My suggestion was going to be to rename any variable that is used by all the tests inside parm/params_grib2_tbl_new, and that does change results.
> 
> My intel and intelllvm tests both completed successfully on Ursa with no baseline changes.

My tests on Ursa confirms the same results as @BenjaminBlake-NOAA 's.

BenjaminBlake-NOAA commented:
My tests with the latest updates completed successfully on WCOSS2, Hera, and Ursa.

I just noticed the new scripts added for Ursa do not have execute permission.  I will do chmod +x on those files to match the permissions of the other scripts in the ci/jobs-dev directory.

Commit:
Add execute permission to Ursa scripts

WenMeng-NOAA commented:
@gspetro-NOAA Can you copy the RT baselines on Ursa from @SamuelTrahanNOAA 's account to EPIC's account?

WenMeng-NOAA commented:
@SamuelTrahanNOAA Do you plan to commit further changes besides switching the RT baseline  to EPIC's directory?

SamuelTrahanNOAA commented:
I plan no further commits to this PR, unless they're requested of me.

gspetro-NOAA commented:
I have rsync'd @SamuelTrahanNOAA 's test_suite directory to `/scratch4/NAGAPE/epic/role-epic/ursa/UPP/test_suite`, so all of the Ursa baselines should be in place there now.

SamuelTrahanNOAA committed:
rt.sh: use EPIC test_suite path on Ursa

SamuelTrahanNOAA commented:
I've updated the path in rt.sh for test_suite on Ursa. The branch should be ready for testing now.

> I have rsync'd SamuelTrahanNOAA 's test_suite directory to /scratch4/NAGAPE/epic/role-epic/ursa/UPP/test_suite, so all of the Ursa baselines should be in place there now.

SamuelTrahanNOAA commented:
@WenMeng-NOAA - Do you want to put author information or documentation in `cmp_grib2_grib2.sh`?

I cleaned it up and put "author: Wen Meng" in it. It's your script, though, so perhaps you should check the documentation and attribution?

SamuelTrahanNOAA reviewed (commented):
It's not normal to review one's own PR, but this one was so complicated I did a line-by-line readthrough anyway. I don't see any problems with it. The `cmp_grib2_grib.sh` comments may need updating, but I'll leave that for its author (Wen Meng) to decide.

Commit:
Add WCOSS2 RT log

BenjaminBlake-NOAA reviewed (approved):
All tests completed successfully on WCOSS2.  @SamuelTrahanNOAA thanks for all your hard work on this.

@gspetro-NOAA This PR is ready for the RTs on R&D machines.  No baseline changes are expected on Hera, Orion, or Hercules.

WenMeng-NOAA commented:
> It's not normal to review one's own PR, but this one was so complicated I did a line-by-line readthrough anyway. I don't see any problems with it. The `cmp_grib2_grib.sh` comments may need updating, but I'll leave that for its author (Wen Meng) to decide.

@SamuelTrahanNOAA Thanks for configurating this script under ci/. It looks good to me. No more update is needed.

WenMeng-NOAA commented:
@SamuelTrahanNOAA Thanks for helping us optimize the UPP RTs.

WenMeng-NOAA commented:
The new RT baselines for Ursa will be created with this PR.

gspetro-NOAA committed:
orion rt log

gspetro-NOAA committed:
hera rt log

gspetro-NOAA committed:
ursa rt logs

gspetro-NOAA committed:
hercules rt log

gspetro-NOAA reviewed (approved):
RTs pass on Hera, Orion, Hercules, and Ursa with no baseline changes.

WenMeng-NOAA commented:
This PR is ready for merging.
